### PR TITLE
RFC: allow Tuple{Union{}}, returning Union{}

### LINKED ISSUE #52385: "Tuple field type cannot be Union{}" error in all 1.10, 1.11, 1.12 versions
I've been running some package tests on the upcoming julia version (1.10rc) and noticed this error newly introduced there, compared to 1.9:
```julia
julia> Tuple{Union{}}
# 1.9:
Tuple{Union{}}
# 1.10:
ERROR: Tuple field type cannot be Union{}
```
Not sure if introduced deliberately or as a side-effect of some other change, but it breaks perfectly working code. And I don't see it anywhere in `Tuple` docs that it shouldn't support all element types.

The above is an MWE, and below is the simplified situation where I actually encountered it:
```julia
julia> X = StructArray(a=Union{}[], b=[]);
# 1.10 throws error right here
# 1.9 lets you proceed just fine

# can work with the non-Union{} column:
julia> X.b
Any[]

# can retrieve the Union{} column as well, of course cannot access its elements
julia> X.a |> typeof
Vector{Union{}} (alias for Array{Union{}, 1})
```

### PR DESCRIPTION
**Pull request by vtjnash**
@oscardssmith requested that I put this up as a PR, where instead of throwing an exception, the call instead returns `Union{}`. It appears this is actually already consistent with the way we had been choosing to represent this particular operation in inference (specific to Tuple, since other types do not use exactly the same code path there), so no changes appear to be necessary there.

 Fixes #52385

### CODE REVIEW

**Review by oscardssmith — APPROVED** [review]

**Comment by ararslan** [comment]
I'd argue that this a bugfix and have marked it for backport to the active release branches accordingly.

**Comment by mbauman** [comment]
My main question is: is this _enough_ of a bug fix?  It doesn't completely fix all of the regressed use-cases identified in #52385.  At a minimum, I think we also need to have `NamedTuple{_, Union{}}` similarly become `Union{}` — both StructArrays and TypedTables need that.

Then the fixes required for StructArrays are two ambiguous methods:

```julia
staticschema(::Type{Union{}}) = Union{}
strip_params(::Type{Union{}}) = identity
```

And the fixes required for TypedTables tests is just an ambiguity/unbound typevar around printing:

```julia
columnnames(::AbstractArray{Union{}}) = ()
columnnames(::Table{<:Any,<:Any,<:NamedTuple{names}}) where {names} = names
``

**Comment by vtjnash** [comment]
Those seem like unambiguously bugfixes regardless of what we do with Tuple{Union{}}?

**Comment by vtjnash** [comment]
Whoops, Oscar and I thought no changes would be necessary in inference with this change, but I found a counter-example today demonstrating that apply_type_tfunc would not be sound after this change:
```julia
julia> const VT = Val{Tuple{T}} where T; code_typed((Type,)) do T; VT{T}; end
1-element Vector{Any}:
 CodeInfo(
    @ REPL[1]:1 within `#3`
1 ─ %1 = Core.apply_type(Main.VT, T)::Type{Val{Tuple{T}}} where T
└──      return %1
) => Type{Val{Tuple{T}}} where T

julia> Val{Union{}} isa Type{Val{Tuple{T}}} where T
false
```

**Comment by vtjnash** [comment]
> How are these "unambiguously bugfixes"

The list of fixes from @mbauman above are equivalent to fixing the output of the ambiguity test checker. I suppose calling them a bugfix is a bit strong though, since it is not actually a bug to have ambiguities, even though their presence causes this test to fail.
```
julia> Test.detect_ambiguities(StructArrays, ambiguous_bottom=true)
```
I believe many packages use `Aqua.test_ambiguities()` to run this test as part of their normal CI and would consider having it passing to be a quality improvement over having it failing.

**Comment by vtjnash** [comment]
This is quite tangential now, but there is actually a performance benefit now to solving those, even though they may seem completely pointless, their presence documents to the compiler that they are pointless, which allows it to skip a lot of work when inferring or loading the package. The added method does need to a slightly peculiar exact form though, with a `Vararg{Any}` slurp argument immediately after the `Type{Union{}}` argument. Due to changes in morespecific matching, this method is also guaranteed to be more specific than any other, so that makes it actually possible to define this without just getting more ambiguities now. The relevant change for that feature addition was: #49470

Secondly, the default ambiguity checker is quite conservative, even though that means it may miss many relevant cases. You can make it less conservative by passing `ambiguous_bottom=true`. A fully non-conservative mode is likely also possible, if there is demand for that. In general currently the algorithm chooses to ignore most cases where ambiguities result from  the heavy use of `Union` or other complex constraints, but we could report those cases as ambiguities instead when it detects them instead of choosing to suppress those when detected. Please file a separate feature request if you would consider that useful and I will see what I can do about it.

**Comment by fingolfin** [comment]
What's the status of this? It has backport labels, but also is marked as a draft / RFC, and has a merge conflict...

**Comment by mbauman** [comment]
What happened here?  I thought that the plan from [#52385](https://github.com/JuliaLang/julia/issues/52385#issuecomment-1969949448) was a good one.  Could you summarize why this isn't the way forward?

**Comment by oscardssmith** [comment]
I believe https://github.com/JuliaLang/julia/pull/54792 is the replacement.
